Change cilium_host IPv6 address

jschwinger233 opened the pull request:
Originally the IPv6 of cilium_host is set to a native address. Instead, this commit uses an IPv6 allocated from IPAM for
it, which is the same way we set IPv4 address for cilium_host.

Fixes: #23445
Fixes: #21954 
Fixes: #23461

```release-note
Change cilium_host IPv6 address, use node router IPv6 instead of native node IPv6, and fixed several relative IPv6 issues.
```

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
pkg: Set cilium_host's IPv6 address to internal router IP

Originally the IPv6 of cilium_host is set to a native address.
Instead, this commit uses an IPv6 allocated from IPAM for
it. This is the same way we set IPv4 address for cilium_host.

To make routing work, this commit also change the `nexthop` for ipv6
routeSpec.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

pippolo84 reviewed (approved):
`/pkg/node` changes LGTM.

I think that the CI failure is related to this test [here](https://github.com/jschwinger233/cilium/blob/e571967dfe78649d3db9c8750483a83095509161/pkg/datapath/linux/node_test.go#L86) not being updated to:

`c.Assert(generatedRoute.Local, checker.DeepEquals, fakeNodeAddressing.IPv6().Router())`

after the change in `createNodeRouteSpec`

jschwinger233 committed:
bpf: remove encap_remap_v6_host_address

The encap_remap_v6_host_address function in encap.h was used to SNAT
egress packets and DNAT ingress packets, replacing HOST_IP by ROUTER_IP
or vice versa. But it's no more needed since we set the cilium_host's
ipv6 to ROUTER_IP.

This commit also removes macro ENABLE_ENCAP_HOST_REMAP.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
pkg: fix broken ipv6 connection when ipsec is on

For traffic from remote host to local pod, the responding skb from local
pod is going through the following steps:

1. lxc_bpf marks the skb with specific encrypt_key and node_id
2. lxc_bpf sends the skb to stack
3. the skb matches a xfrm policy and proceeds with the ipsec encap
4. the skb encapped by ipsec goes with another specific mark
5. the skb traverses the remaining stack and goes out to eth0

There were two bugs in this datapath for ipv6:

1. ipv6 xfrm policy was wrongly configured:

The correct xfrm policy should be like:

src $cidr_local dst $cidr_remote
	dir out priority 0
	mark $mark
	tmpl src $ipv6_local dst $ipv6_remote
		proto esp spi $seq reqid 1 mode tunnel

However, we were using local router ip in the "src", rather than local
pod cidr. This mistake prevented return skb from being ipsec-encapped.

2. ip6tables was wrongly set

Looking at the forementioned steps, the ipsec-encapped skb shouldn't
have any other NAT but go directly to eth0; but the current ip6tables
would do the masquerading and mess up the skb.

This commit fixed both bugs.

For the first one, I simply revised the code where we set xfrm policy
and use the local pod cidr replacing local router ip.

For the second one, I add some extra ip6tables rules for excluding
ipsec-encapped skb, which can be done by matching mark. Ip4tables has
done the same things for ipv4, and I just applied these rules for ipv6.

Please be aware that so far we don't have testcase to cover ipv6 with
ipsec on, so this fix is tested manually.

Fixes: #23461

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

brb reviewed (approved):
:exploding_head: amazing!

jschwinger233 committed:
pkg: fix ip6tables to properly handle proxy return traffic

Within a cluster without KPR disabled, an ipv6 traffic from host1 to
nodeport-svc@host2 is expected to go through following steps on host2:

1. DNAT by ip6tables chain KUBE-NODEPORTS
2. routing decision: cilium_host
3. MASQUERADE by ip6tables chain KUBE-POSTROUTING
4. tc egress filter on cilium_host: set skb mark for further tproxy
5. ingress from cilium_net
6. traverse ip6tables again, TPROXY-ed by chain CILIUM_PRE_mangle

Of course, a return traffic is supposed to be rev-NAT-ed by conntrack,
but at the moment the following ip6tables rule prevents that:

-A CILIUM_OUTPUT_raw -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack

This commits changed this rule and aligned with ip4tables to make
conntrack work:

-A CILIUM_OUTPUT_raw -o lxc+ -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o cilium_host -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack

After this patch, the return traffic routed to eth0 or lo will do the
rev-NAT as we wish, so as to fix the issue

Fixes: #21954

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1611,31 +1611,36 @@ func (m *IptablesManager) addCiliumAcceptXfrmRules() error {
-		return err
+
+	for _, insertOpt := range insertOpts {
+		if err := insertAcceptXfrm(ip4tables, insertOpt.table, insertOpt.chain); err != nil {
```
Should we skip this if IPv4 is disabled?

pchaigno reviewed (commented):
Nice work!

If we are fixing IPsec+IPv6, shouldn't we be able to reenable some tests to cover it?

pchaigno reviewed (approved):
> If we are fixing IPsec+IPv6, shouldn't we be able to reenable some tests to cover it?

That can probably be a follow up PR if we don't want to delay merging this one.

pchaigno reviewed (approved):
:heart: for the improved commit descriptions.

julianwiedmann commented on bpf/Makefile:
```diff
@@ -58,7 +58,7 @@ LB_OPTIONS = \
 	-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC: \
-	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DENABLE_ENCAP_HOST_REMAP: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC: \
```
Once merged, could you open a follow-on PR to remove this duplication? No need to fix it here unless you need to push for some other reason :+1:.

jschwinger233 replied:
For sure! I took 30 seconds to understand what duplication you were saying LOL
